File: Sources/KiraCompiler/Parser/AST.swift
import Foundation

public struct Annotation: Sendable {
    public struct Argument: Sendable {
        public var label: String?
        public var value: Expr

        public init(label: String?, value: Expr) {
            self.label = label
            self.value = value
        }
    }

    public var name: String
    public var arguments: [Argument]
    public var range: SourceRange

    public init(name: String, arguments: [Argument], range: SourceRange) {
        self.name = name
        self.arguments = arguments
        self.range = range
    }
}

public struct TypeRef: Sendable {
    public indirect enum Kind: Sendable {
        case named(String)
        case applied(String, [TypeRef])
        case fixedArray(element: TypeRef, count: Int)
        case array(TypeRef)
        case dictionary(key: TypeRef, value: TypeRef)
        case optional(TypeRef)
        case function(params: [TypeRef], returns: TypeRef)
    }

    public var kind: Kind
    public var range: SourceRange

    public init(kind: Kind, range: SourceRange) {
        self.kind = kind
        self.range = range
    }
}

public struct ModuleAST: Sendable {
    public var imports: [ImportDecl]
    public var declarations: [Decl]
    public var range: SourceRange

    public init(imports: [ImportDecl], declarations: [Decl], range: SourceRange) {
        self.imports = imports
        self.declarations = declarations
        self.range = range
    }
}

public struct ImportDecl: Sendable {
    public var modulePath: String
    public var alias: String?
    public var range: SourceRange

    public init(modulePath: String, alias: String?, range: SourceRange) {
        self.modulePath = modulePath
        self.alias = alias
        self.range = range
    }
}

public enum Decl: Sendable {
    case function(FunctionDecl)
    case externFunction(ExternFunctionDecl)
    case `typealias`(TypealiasDecl)
    case type(TypeDecl)
    case `protocol`(ProtocolDecl)
    case `enum`(EnumDecl)
    case construct(ConstructDecl)
    case constructInstance(ConstructInstanceDecl)
    case globalVar(VarDeclStmt)

    public var range: SourceRange {
        switch self {
        case .function(let decl): return decl.range
        case .externFunction(let decl): return decl.range
        case .typealias(let decl): return decl.range
        case .type(let decl): return decl.range
        case .protocol(let decl): return decl.range
        case .enum(let decl): return decl.range
        case .construct(let decl): return decl.range
        case .constructInstance(let decl): return decl.range
        case .globalVar(let decl): return decl.range
        }
    }
}

public struct TypealiasDecl: Sendable {
    public var name: String
    public var target: TypeRef
    public var range: SourceRange

    public init(name: String, target: TypeRef, range: SourceRange) {
        self.name = name
        self.target = target
        self.range = range
    }
}

public struct Parameter: Sendable {
    public var name: String
    public var type: TypeRef
    public var range: SourceRange

    public init(name: String, type: TypeRef, range: SourceRange) {
        self.name = name
        self.type = type
        self.range = range
    }
}

public struct FunctionDecl: Sendable {
    public var annotations: [Annotation]
    public var isAsync: Bool
    public var name: String
    public var parameters: [Parameter]
    public var returnType: TypeRef?
    public var body: BlockStmt
    public var range: SourceRange

    public init(annotations: [Annotation], isAsync: Bool, name: String, parameters: [Parameter], returnType: TypeRef?, body: BlockStmt, range: SourceRange) {
        self.annotations = annotations
        self.isAsync = isAsync
        self.name = name
        self.parameters = parameters
        self.returnType = returnType
        self.body = body
        self.range = range
    }
}

public struct ExternFunctionDecl: Sendable {
    public var annotations: [Annotation]
    public var name: String
    public var parameters: [Parameter]
    public var returnType: TypeRef?
    public var range: SourceRange

    public init(annotations: [Annotation], name: String, parameters: [Parameter], returnType: TypeRef?, range: SourceRange) {
        self.annotations = annotations
        self.name = name
        self.parameters = parameters
        self.returnType = returnType
        self.range = range
    }
}

public struct TypeDecl: Sendable {
    public struct Field: Sendable {
        public var annotations: [Annotation]
        public var isStatic: Bool
        public var isVar: Bool
        public var name: String
        public var type: TypeRef?
        public var initializer: Expr?
        public var range: SourceRange

        public init(annotations: [Annotation], isStatic: Bool, isVar: Bool, name: String, type: TypeRef?, initializer: Expr?, range: SourceRange) {
            self.annotations = annotations
            self.isStatic = isStatic
            self.isVar = isVar
            self.name = name
            self.type = type
            self.initializer = initializer
            self.range = range
        }
    }

    public var annotations: [Annotation]
    public var name: String
    public var conformances: [String]
    public var fields: [Field]
    public var methods: [FunctionDecl]
    public var statics: [FunctionDecl]
    public var range: SourceRange

    public init(annotations: [Annotation], name: String, conformances: [String], fields: [Field], methods: [FunctionDecl], statics: [FunctionDecl], range: SourceRange) {
        self.annotations = annotations
        self.name = name
        self.conformances = conformances
        self.fields = fields
        self.methods = methods
        self.statics = statics
        self.range = range
    }
}

public struct ProtocolDecl: Sendable {
    public struct Requirement: Sendable {
        public var annotations: [Annotation]
        public var name: String
        public var parameters: [Parameter]
        public var returnType: TypeRef?
        public var range: SourceRange
    }

    public var annotations: [Annotation]
    public var name: String
    public var requirements: [Requirement]
    public var range: SourceRange
}

public struct EnumDecl: Sendable {
    public struct Case: Sendable {
        public struct AssociatedValue: Sendable {
            public var label: String?
            public var type: TypeRef
            public var range: SourceRange
        }
        public var annotations: [Annotation]
        public var name: String
        public var associatedValues: [AssociatedValue]
        public var range: SourceRange
    }

    public var annotations: [Annotation]
    public var name: String
    public var cases: [Case]
    public var range: SourceRange
}

public struct ConstructDecl: Sendable {
    public struct Modifier: Sendable {
        public var isScoped: Bool
        public var name: String
        public var type: TypeRef?
        public var defaultValue: Expr?
        public var range: SourceRange
    }

    public var name: String
    public var allowedAnnotations: [String]
    public var modifiers: [Modifier]
    public var requiredBlocks: [String]
    public var range: SourceRange
}

public struct ConstructInstanceDecl: Sendable {
    public struct Member: Sendable {
        public enum Kind: Sendable {
            case field(TypeDecl.Field)
            case function(FunctionDecl)
            case block(name: String, body: BlockStmt)
        }
        public var kind: Kind
        public var range: SourceRange
    }

    public var annotations: [Annotation]
    public var constructName: String
    public var name: String
    public var parameters: [Parameter]
    public var members: [Member]
    public var range: SourceRange
}

public struct BlockStmt: Sendable {
    public var statements: [Stmt]
    public var range: SourceRange

    public init(statements: [Stmt], range: SourceRange) {
        self.statements = statements
        self.range = range
    }
}

public enum Stmt: Sendable {
    case variable(VarDeclStmt)
    case `return`(ReturnStmt)
    case expr(Expr)
    case `if`(IfStmt)
    case `while`(WhileStmt)
    case match(MatchStmt)

    public var range: SourceRange {
        switch self {
        case .variable(let stmt): return stmt.range
        case .return(let stmt): return stmt.range
        case .expr(let expr): return expr.range
        case .if(let stmt): return stmt.range
        case .while(let stmt): return stmt.range
        case .match(let stmt): return stmt.range
        }
    }
}

public struct VarDeclStmt: Sendable {
    public var isVar: Bool
    public var name: String
    public var explicitType: TypeRef?
    public var initializer: Expr
    public var range: SourceRange
}

public struct ReturnStmt: Sendable {
    public var value: Expr?
    public var range: SourceRange
}

public struct IfStmt: Sendable {
    public var condition: Expr
    public var thenBlock: BlockStmt
    public var elseBlock: BlockStmt?
    public var range: SourceRange
}

public struct WhileStmt: Sendable {
    public var condition: Expr
    public var body: BlockStmt
    public var range: SourceRange
}

public struct MatchStmt: Sendable {
    public struct Pattern: Sendable {
        public var variantName: String
        public var bindings: [String]
        public var range: SourceRange
    }

    public struct Case: Sendable {
        public var pattern: Pattern
        public var body: BlockStmt
        public var range: SourceRange
    }

    public var value: Expr
    public var cases: [Case]
    public var range: SourceRange
}

public indirect enum Expr: Sendable {
    case identifier(String, SourceRange)
    case leadingMember(String, SourceRange)
    case intLiteral(Int64, SourceRange)
    case floatLiteral(Double, SourceRange)
    case stringLiteral(String, SourceRange)
    case boolLiteral(Bool, SourceRange)
    case nilLiteral(SourceRange)
    case sizeOf(SizeOfExpr)
    case arrayLiteral(ArrayLiteralExpr)
    case unary(UnaryExpr)
    case binary(BinaryExpr)
    case conditional(ConditionalExpr)
    case call(CallExpr)
    case member(MemberExpr)
    case index(IndexExpr)
    case assign(AssignExpr)
    case shaderMacro(ShaderMacroExpr)

    public var range: SourceRange {
        switch self {
        case .identifier(_, let r): return r
        case .leadingMember(_, let r): return r
        case .intLiteral(_, let r): return r
        case .floatLiteral(_, let r): return r
        case .stringLiteral(_, let r): return r
        case .boolLiteral(_, let r): return r
        case .nilLiteral(let r): return r
        case .sizeOf(let s): return s.range
        case .arrayLiteral(let a): return a.range
        case .unary(let u): return u.range
        case .binary(let b): return b.range
        case .conditional(let c): return c.range
        case .call(let c): return c.range
        case .member(let m): return m.range
        case .index(let i): return i.range
        case .assign(let a): return a.range
        case .shaderMacro(let s): return s.range
        }
    }
}

public struct SizeOfExpr: Sendable {
    public var type: TypeRef
    public var range: SourceRange
    public init(type: TypeRef, range: SourceRange) {
        self.type = type
        self.range = range
    }
}

public struct ArrayLiteralExpr: Sendable {
    public var elements: [Expr]
    public var range: SourceRange
    public init(elements: [Expr], range: SourceRange) {
        self.elements = elements
        self.range = range
    }
}

public struct UnaryExpr: Sendable {
    public var op: UnaryOp
    public var expr: Expr
    public var range: SourceRange
    public init(op: UnaryOp, expr: Expr, range: SourceRange) {
        self.op = op
        self.expr = expr
        self.range = range
    }
}

public struct BinaryExpr: Sendable {
    public var op: BinaryOp
    public var lhs: Expr
    public var rhs: Expr
    public var range: SourceRange
    public init(op: BinaryOp, lhs: Expr, rhs: Expr, range: SourceRange) {
        self.op = op
        self.lhs = lhs
        self.rhs = rhs
        self.range = range
    }
}

public struct ConditionalExpr: Sendable {
    public var condition: Expr
    public var thenExpr: Expr
    public var elseExpr: Expr
    public var range: SourceRange
    public init(condition: Expr, thenExpr: Expr, elseExpr: Expr, range: SourceRange) {
        self.condition = condition
        self.thenExpr = thenExpr
        self.elseExpr = elseExpr
        self.range = range
    }
}

public struct CallExpr: Sendable {
    public var callee: Expr
    public var arguments: [CallArgument]
    public var trailingBlock: BlockStmt?
    public var range: SourceRange
    public init(callee: Expr, arguments: [CallArgument], trailingBlock: BlockStmt?, range: SourceRange) {
        self.callee = callee
        self.arguments = arguments
        self.trailingBlock = trailingBlock
        self.range = range
    }
}

public struct MemberExpr: Sendable {
    public var base: Expr
    public var name: String
    public var range: SourceRange
    public init(base: Expr, name: String, range: SourceRange) {
        self.base = base
        self.name = name
        self.range = range
    }
}

public struct IndexExpr: Sendable {
    public var base: Expr
    public var index: Expr
    public var range: SourceRange
    public init(base: Expr, index: Expr, range: SourceRange) {
        self.base = base
        self.index = index
        self.range = range
    }
}

public struct AssignExpr: Sendable {
    public var target: Expr
    public var value: Expr
    public var range: SourceRange
    public init(target: Expr, value: Expr, range: SourceRange) {
        self.target = target
        self.value = value
        self.range = range
    }
}

public struct ShaderMacroExpr: Sendable {
    public var functionName: String
    public var range: SourceRange
    public init(functionName: String, range: SourceRange) {
        self.functionName = functionName
        self.range = range
    }
}

public struct CallArgument: Sendable {
    public var label: String?
    public var value: Expr
    public var range: SourceRange
}

public enum UnaryOp: Sendable { case negate, not }

public enum BinaryOp: Sendable {
    case add, sub, mul, div, mod
    case eq, neq, lt, gt, lte, gte
    case and, or
}

File: packages/kira_syntax_model/src/ast.zig
const std = @import("std");
const Span = @import("kira_source").Span;

pub const Program = struct {
    imports: []ImportDecl,
    decls: []Decl,
    functions: []FunctionDecl,
};

pub const Decl = union(enum) {
    annotation_decl: AnnotationDecl,
    capability_decl: CapabilityDecl,
    enum_decl: EnumDecl,
    function_decl: FunctionDecl,
    type_decl: TypeDecl,
    construct_decl: ConstructDecl,
    construct_form_decl: ConstructFormDecl,
};

pub const ImportDecl = struct {
    module_name: QualifiedName,
    alias: ?[]const u8,
    span: Span,
};

pub const NameSegment = struct {
    text: []const u8,
    span: Span,
};

pub const QualifiedName = struct {
    segments: []NameSegment,
    span: Span,
};

pub const Annotation = struct {
    name: QualifiedName,
    args: []AnnotationArg,
    block: ?AnnotationBlock,
    span: Span,
};

pub const AnnotationArg = struct {
    label: ?[]const u8,
    value: *Expr,
    span: Span,
};

pub const AnnotationBlock = struct {
    entries: []AnnotationBlockEntry,
    span: Span,
};

pub const AnnotationBlockEntry = union(enum) {
    value: AnnotationBlockValue,
    field: AnnotationBlockField,
};

pub const AnnotationBlockValue = struct {
    value: *Expr,
    span: Span,
};

pub const AnnotationBlockField = struct {
    name: []const u8,
    value: *Expr,
    span: Span,
};

pub const AnnotationDecl = struct {
    name: []const u8,
    targets: []AnnotationTarget,
    uses: []QualifiedName,
    parameters: []AnnotationParameterDecl,
    generated_members: []GeneratedMember,
    span: Span,
};

pub const CapabilityDecl = struct {
    name: []const u8,
    generated_members: []GeneratedMember,
    span: Span,
};

pub const AnnotationTarget = enum {
    class,
    struct_decl,
    function,
    construct,
    field,
};

pub const AnnotationParameterDecl = struct {
    name: []const u8,
    type_expr: *TypeExpr,
    default_value: ?*Expr,
    span: Span,
};

pub const GeneratedMember = struct {
    overridable: bool,
    member: BodyMember,
    span: Span,
};

pub const FunctionDecl = struct {
    annotations: []const Annotation,
    is_override: bool = false,
    name: []const u8,
    params: []ParamDecl,
    return_type: ?*TypeExpr,
    body: ?Block,
    span: Span,
};

pub const FunctionSignature = struct {
    name: []const u8,
    params: []ParamDecl,
    return_type: ?*TypeExpr,
    span: Span,
};

pub const ParamDecl = struct {
    annotations: []const Annotation,
    name: []const u8,
    type_expr: ?*TypeExpr,
    span: Span,
};

pub const EnumDecl = struct {
    name: []const u8,
    type_params: [][]const u8,
    variants: []EnumVariantDecl,
    span: Span,
};

pub const EnumVariantDecl = struct {
    name: []const u8,
    associated_type: ?*TypeExpr,
    default_value: ?*Expr,
    span: Span,
};

pub const TypeDecl = struct {
    kind: TypeKind,
    annotations: []const Annotation,
    name: []const u8,
    parents: []QualifiedName,
    members: []BodyMember,
    span: Span,
};

pub const TypeKind = enum {
    class,
    struct_decl,
};

pub const ConstructDecl = struct {
    annotations: []const Annotation,
    name: []const u8,
    sections: []ConstructSection,
    span: Span,
};

pub const ConstructSection = struct {
    name: []const u8,
    kind: ConstructSectionKind,
    entries: []ConstructSectionEntry,
    span: Span,
};

pub const ConstructSectionKind = enum {
    annotations,
    modifiers,
    requires,
    lifecycle,
    builder,
    representation,
    custom,
};

pub const ConstructSectionEntry = union(enum) {
    annotation_spec: AnnotationSpec,
    field_decl: FieldDecl,
    lifecycle_hook: LifecycleHook,
    function_signature: FunctionSignature,
    named_rule: NamedRule,
};

pub const AnnotationSpec = struct {
    name: QualifiedName,
    type_expr: ?*TypeExpr,
    default_value: ?*Expr,
    span: Span,
};

pub const NamedRule = struct {
    name: QualifiedName,
    args: []RuleArg,
    type_expr: ?*TypeExpr,
    value: ?*Expr,
    block: ?Block,
    span: Span,
};

pub const RuleArg = struct {
    label: ?[]const u8,
    value: ?*Expr,
    span: Span,
};

pub const ConstructFormDecl = struct {
    annotations: []const Annotation,
    construct_name: QualifiedName,
    name: []const u8,
    params: []ParamDecl,
    body: ConstructBody,
    span: Span,
};

pub const ConstructBody = struct {
    members: []BodyMember,
    span: Span,
};

pub const BodyMember = union(enum) {
    field_decl: FieldDecl,
    function_decl: FunctionDecl,
    content_section: ContentSection,
    lifecycle_hook: LifecycleHook,
    named_rule: NamedRule,
};

pub const FieldDecl = struct {
    annotations: []const Annotation,
    is_override: bool = false,
    storage: FieldStorage,
    name: []const u8,
    type_expr: ?*TypeExpr,
    value: ?*Expr,
    span: Span,
};

pub const FieldStorage = enum {
    immutable,
    mutable,
};

pub const ContentSection = struct {
    annotations: []const Annotation,
    builder: BuilderBlock,
    span: Span,
};

pub const LifecycleHook = struct {
    name: []const u8,
    args: []RuleArg,
    body: Block,
    span: Span,
};

pub const Block = struct {
    statements: []Statement,
    span: Span,
};

pub const Statement = union(enum) {
    let_stmt: LetStatement,
    assign_stmt: AssignStatement,
    expr_stmt: ExprStatement,
    return_stmt: ReturnStatement,
    if_stmt: IfStatement,
    for_stmt: ForStatement,
    while_stmt: WhileStatement,
    break_stmt: BreakStatement,
    continue_stmt: ContinueStatement,
    match_stmt: MatchStatement,
    switch_stmt: SwitchStatement,
};

pub const LetStatement = struct {
    annotations: []const Annotation,
    storage: FieldStorage,
    name: []const u8,
    type_expr: ?*TypeExpr,
    value: ?*Expr,
    span: Span,
};

pub const ExprStatement = struct {
    expr: *Expr,
    span: Span,
};

pub const AssignStatement = struct {
    target: *Expr,
    value: *Expr,
    span: Span,
};

pub const ReturnStatement = struct {
    value: ?*Expr,
    span: Span,
};

pub const IfStatement = struct {
    condition: *Expr,
    then_block: Block,
    else_block: ?Block,
    span: Span,
};

pub const ForStatement = struct {
    binding_name: []const u8,
    iterator: *Expr,
    body: Block,
    span: Span,
};

pub const WhileStatement = struct {
    condition: *Expr,
    body: Block,
    span: Span,
};

pub const BreakStatement = struct {
    span: Span,
};

pub const ContinueStatement = struct {
    span: Span,
};

pub const MatchStatement = struct {
    subject: *Expr,
    arms: []MatchArm,
    span: Span,
};

pub const MatchArm = struct {
    patterns: []MatchPattern,
    guard: ?*Expr,
    body: Block,
    span: Span,
};

pub const MatchPattern = union(enum) {
    bare_variant: struct { name: []const u8, span: Span },
    destructure: struct { variant_name: []const u8, inner: *MatchPattern, span: Span },
    as_binding: struct { inner: *MatchPattern, binding_name: []const u8, span: Span },
};

pub const SwitchStatement = struct {
    subject: *Expr,
    cases: []SwitchCase,
    default_block: ?Block,
    span: Span,
};

pub const SwitchCase = struct {
    pattern: *Expr,
    body: Block,
    span: Span,
};

pub const BuilderBlock = struct {
    items: []BuilderItem,
    span: Span,
};

pub const BuilderItem = union(enum) {
    expr: BuilderExprItem,
    if_item: BuilderIfItem,
    for_item: BuilderForItem,
    switch_item: BuilderSwitchItem,
};

pub const BuilderExprItem = struct {
    expr: *Expr,
    span: Span,
};

pub const BuilderIfItem = struct {
    condition: *Expr,
    then_block: BuilderBlock,
    else_block: ?BuilderBlock,
    span: Span,
};

pub const BuilderForItem = struct {
    binding_name: []const u8,
    iterator: *Expr,
    body: BuilderBlock,
    span: Span,
};

pub const BuilderSwitchItem = struct {
    subject: *Expr,
    cases: []BuilderSwitchCase,
    default_block: ?BuilderBlock,
    span: Span,
};

pub const BuilderSwitchCase = struct {
    pattern: *Expr,
    body: BuilderBlock,
    span: Span,
};

pub const Expr = union(enum) {
    integer: IntegerLiteral,
    float: FloatLiteral,
    string: StringLiteral,
    bool: BoolLiteral,
    identifier: IdentifierExpr,
    array: ArrayExpr,
    callback: CallbackBlock,
    struct_literal: StructLiteralExpr,
    native_state: NativeStateExpr,
    native_user_data: NativeUserDataExpr,
    native_recover: NativeRecoverExpr,
    unary: UnaryExpr,
    binary: BinaryExpr,
    conditional: ConditionalExpr,
    member: MemberExpr,
    index: IndexExpr,
    call: CallExpr,
};

pub const IntegerLiteral = struct {
    value: i64,
    span: Span,
};

pub const FloatLiteral = struct {
    value: f64,
    span: Span,
};

pub const StringLiteral = struct {
    value: []const u8,
    span: Span,
};

pub const BoolLiteral = struct {
    value: bool,
    span: Span,
};

pub const IdentifierExpr = struct {
    name: QualifiedName,
    span: Span,
};

pub const ArrayExpr = struct {
    elements: []*Expr,
    span: Span,
};

pub const StructLiteralExpr = struct {
    type_name: QualifiedName,
    fields: []StructLiteralField,
    span: Span,
};

pub const StructLiteralField = struct {
    name: []const u8,
    value: *Expr,
    span: Span,
};

pub const NativeStateExpr = struct {
    value: *Expr,
    span: Span,
};

pub const NativeUserDataExpr = struct {
    state: *Expr,
    span: Span,
};

pub const NativeRecoverExpr = struct {
    state_type: *TypeExpr,
    value: *Expr,
    span: Span,
};

pub const UnaryExpr = struct {
    op: UnaryOp,
    operand: *Expr,
    span: Span,
};

pub const BinaryExpr = struct {
    op: BinaryOp,
    lhs: *Expr,
    rhs: *Expr,
    span: Span,
};

pub const ConditionalExpr = struct {
    condition: *Expr,
    then_expr: *Expr,
    else_expr: *Expr,
    span: Span,
};

pub const MemberExpr = struct {
    object: *Expr,
    member: []const u8,
    span: Span,
};

pub const IndexExpr = struct {
    object: *Expr,
    index: *Expr,
    span: Span,
};

pub const CallExpr = struct {
    callee: *Expr,
    args: []CallArg,
    trailing_builder: ?BuilderBlock,
    trailing_callback: ?CallbackBlock,
    span: Span,
};

pub const CallArg = struct {
    label: ?[]const u8,
    value: *Expr,
    span: Span,
};

pub const BinaryOp = enum {
    add,
    subtract,
    multiply,
    divide,
    modulo,
    equal,
    not_equal,
    less,
    less_equal,
    greater,
    greater_equal,
    logical_and,
    logical_or,
};

pub const UnaryOp = enum {
    negate,
    not,
};

pub const TypeExpr = union(enum) {
    named: QualifiedName,
    generic: GenericTypeExpr,
    any: AnyTypeExpr,
    array: ArrayTypeExpr,
    function: FunctionTypeExpr,
};

pub const AnyTypeExpr = struct {
    target: *TypeExpr,
    span: Span,
};

pub const ArrayTypeExpr = struct {
    element_type: *TypeExpr,
    span: Span,
};

pub const FunctionTypeExpr = struct {
    params: []*TypeExpr,
    result: *TypeExpr,
    span: Span,
};

pub const GenericTypeExpr = struct {
    base: QualifiedName,
    args: []*TypeExpr,
    span: Span,
};

pub const CallbackBlock = struct {
    params: []CallbackParam,
    body: Block,
    span: Span,
};

pub const CallbackParam = struct {
    name: []const u8,
    span: Span,
};

pub const dumpProgram = @import("ast_dump.zig").dumpProgram;
